src/python/pants/backend/java/target_types.py

```diff
@@ -18,40 +18,68 @@
 from pants.engine.unions import UnionMembership, UnionRule
 
 
-class JavaSources(Sources):
+class JavaSourceField(Sources):
     expected_file_extensions = (".java",)
+    expected_num_files = 1
+
+
+class JavaGeneratorSources(Sources):
+    pass
+    # TODO: do I need anything else here?
+
+
+# -----------------------------------------------------------------------------------------------
+# `java_test` target
+# -----------------------------------------------------------------------------------------------
+
+
+class JavaTestSourceField(JavaSourceField):
+    pass
+
+
+class JunitTestTarget(Target):
+    alias = "junit_test"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Dependencies,
+        JavaTestSourceField,
+    )
+    help = "A single Java test, run with JUnit."
 
 
 # -----------------------------------------------------------------------------------------------
-# `java_tests` target
+# `junit_tests` target generator
 # -----------------------------------------------------------------------------------------------
 
 
-class JavaTestsSources(JavaSources):
+class JavaTestsGeneratorSourcesField(JavaGeneratorSources):
     default = ("*Test.java",)
 
 
-class JunitTests(Target):
+class JunitTestsGeneratorTarget(Target):
     alias = "junit_tests"
     core_fields = (
         *COMMON_TARGET_FIELDS,
+        JavaTestsGeneratorSourcesField,
         Dependencies,
-        JavaTestsSources,
     )
-    help = "Java tests, run with Junit."
+    help = "A suite of Java tests, run with JUnit."
```

Comment:
This isn't quite accurate. I think probably want language more like this:

https://github.com/pantsbuild/pants/blob/55e93bbb8defa06d5a055c9119056d4419a2cfdf/src/python/pants/backend/terraform/target_types.py#L36-L37

You can probably skip the "where Java files are present" part.

Reply:
Is that necessarily correct? Aren't we generating a target for each file, rather than each directory?

Reply:
Indeed, that part needs to be changed. I meant something like that, not the exact thing

Reply:
Makes sense. Thanks :)